pilot/pkg/serviceregistry/kube/controller/endpointslice.go

```diff
@@ -359,7 +359,7 @@ func (esc *endpointSliceController) InstancesByPort(c *Controller, svc *model.Se
 func (esc *endpointSliceController) newEndpointBuilder(pod *corev1.Pod) *EndpointBuilder {
 	if pod != nil {
 		// Respect pod "istio-locality" label
-		if pod.Labels[model.LocalityLabel] == "" {
+		if pod.Labels != nil && pod.Labels[model.LocalityLabel] == "" {
```

Comment:
If I'm reading the code correctly, `NewEndpointBuilder()` doesn't set the `"istio-locality"` label (`model.LocalityLabel`). It takes the existing labels and passes them to `AugmentLabels()` along with result from `getPodLocality()`, like you pointed out, then splits up that locality into a variety of other labels that is then returned.

So yea, it will have parts of the locality as other labels on the pod, but it doesn't specifically set `"istio-locality"`, which is probably why that logic exists. Maybe it would be better to update `AugmentLabels()` to add the `"istio-locality"` label than to have it in `NewEndpointBuilder()`? Since that's where label munging is happening anyways?

Reply:
I would agree with you - it is technically changing the behavior. However, this behavior is not present in Endpoints mode (for context, we can read Endpoints _or_ EndpointSlice), and I am fairly confident it is accidental. Thanks for digging deep into this.

So I would actually remove this, which will result in `istio-locality` NOT being set, which is ok.